packages/plugin-kafka/__tests__/kafka.spec.ts

```diff
@@ -0,0 +1,197 @@
+    producerStub = stub(Producer.prototype, 'connect')
+
+    function fakeConnect(_?: any, cb?: ((err: any, data: any) => any)) {
+      cb!(new RdKafkaError(-195), {})
+    }
+
+    // @ts-ignore
+    consumerStub.callsFake(fakeConnect)
+    // @ts-ignore
+    producerStub.callsFake(fakeConnect)
+  })
+
+  afterAll(() => {
+    consumerStub.restore()
+    producerStub.restore()
+  })
+
+  test('should be able to create a producer', async () => {
+    const createPromise = service.kafka.createProducerStream({ objectMode: false, topic: 'testBoo' })
+    await expect(createPromise).rejects.toThrow(/Broker transport failure/)
+  })
+
+  test('should be able to create a consumer', async () => {
+    const createPromise = service.kafka.createConsumerStream({ topics: ['test'], connectOptions: { timeout: 1000 } },  { 'client.id': 'consume-group' })
+    await expect(createPromise).rejects.toThrow(/Broker transport failure/)
+  })
+
+})
+
+describe('connected to broker', () => {
+  function getMessageIterable(count: number) {
+    const sentMessages: any[] = []
+    function *messagesToSend(topic: string) {
+      for (let i = 0; i < count; i += 1) {
+        const message =  {
+          topic,
+          value: Buffer.from(`message ${i} at ${Date.now()}`),
+        }
+        sentMessages.push(message)
+        yield message
+      }
+    }
+
+    return {
+      sentMessages,
+      messagesToSend,
+    }
+  }
+
+  test('consume/produce no-auto-commit', async () => {
+    const topic = 'test-no-auto-consume-produce'
+    const messagesToPublish = 5
+
+    const messageIterable = getMessageIterable(messagesToPublish)
+    const messageStream = Readable.from(messageIterable.messagesToSend(topic), { autoDestroy: true })
+
+    producer = await service.kafka.createProducerStream({ objectMode: true, pollInterval: 10 }, {
+      'group.id': 'other-group',
+    })
+
+    consumer = await service.kafka.createConsumerStream(
+      { topics: topic, streamAsBatch: true, fetchSize: 10 },
+      {
+        debug: 'consumer',
+        'auto.commit.enable': false,
+        'client.id': 'someid',
+        'group.id': 'other-group',
+      },
+      {
+        'auto.offset.reset': 'earliest',
+      }
+    )
+
+    const pipelinePromise = promisify(pipeline)
+    const receivedMessages:any[] = []
+
+    const consumeFn = async () => {
+      for await (const incommingMessage of consumer) {
+        const messages = Array.isArray(incommingMessage) ? incommingMessage : [incommingMessage]
+        receivedMessages.push(...messages)
+        consumer.consumer.commitMessageSync(messages.pop())
+        if (receivedMessages.length >= messagesToPublish) {
+          await promisify(consumer.close.bind(consumer))()
+        }
+      }
+    }
+
+    await pipelinePromise(messageStream, producer)
+
+    await Promise.race([
+      consumeFn(),
+      new Promise(resolve => setTimeout(() => { consumer.close(); resolve() }, 7000)),
+    ])
```

Comment:
this part seems messy. any thoughts on how we can improve it?
promisifying bound function & immediately calling it seems excessive

Reply:
Tried to make it look cleaner